### PR TITLE
Hifitime blog post

### PR DESCRIPTION
**Pull request by ChristopherRabotin**
Hi there!

Here is my first decent draft of the blog post for Kani. I can definitely expand more on the tests cases that were solved and how if you wish. I felt the need to explain why hifitime is even relevant to be formally tested: let me know if that part is too long.

Best,
Chris

### CODE REVIEW

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
...00 seconds regardless of whether it's 01:00 or 14:00. If we only follow this fixed definition of an hour, then time drifts because the Earth does not in fact complete a full rotation on itself in _exactly_ 23 hours 56 minutes and 4.091 seconds (that's the duration of a sidereal day). For "human time" (Universal Coordinated Time, `UTC`) to catch up with the time with respect to the stars (`UT1`), we regularly introduce "leap seconds." Like leap years, where we introduce an extra day every fourth year (with rare exceptions), leap seconds introduce an extra seconds _every so often_, as [announced by the IERS](https://www.ietf.org/timezones/data/leap-seconds.list). When scientists report to the IERS that the Earth is about to drift quite a bit compared to `UTC`, IERS announces that on a given day at least six months in the future, there will be an extra second in that day to allow for the rotation of the Earth to catch up. In practice, this typically means that UTC clocks "stop" the time for one second: this means UTC is _not_ a continuous time scale: that second we didn't count in UTC still happened for the universe! As the Standards Of Fundamental Astronomy (SOFA) eloquently puts it:
```
Comment (tautschnig) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:16]: s/extra seconds/extra second/ (before "_every so often_").

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+> -- "SOFA Time Scales and Calendar Tools", Document version 1.61, section 3.5.1
+
+<img src="{{site.baseurl | prepend: site.url}}/assets/images/utc-time-scale-drift.png" alt="UTC drifts with respect to other time scales" />
+
+Luckily there a few glitch-free time scales to choose from. Hifitime uses TAI, the international atomic time. This allows quick synchronization with UTC time because both UTC and TAI "tick" at the same rate as they are both Earth based time scales. In fact, with general relativity, we understand that time is affected by gravity fields. This causes the additional problem that a second does not tick with the same frequency on Earth as it does in the vacuum between Mars and Jupiter (which is more than twice as far apart as Mars is from the Sun by the way). As such, the position of planets published by NASA JPL are provided in a time scale in which the influence of the gravity of Earth has been removed and where the "ticks" of one second are fixed at the solar system barycenter: the Dynamic Barycentric Time (TDB) time scale.
```
Comment (tautschnig) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:23]: "Luckily there are a few" (missing "are")

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+Hence, to avoid loss of precision, hifitime stores all durations as a tuple of the number of centuries and the number of nanoseconds into this century. For example `(0, 1)` is a duration of one nanosecond. Then, the Epoch is a structure containing the TAI duration since the TAI reference epoch of 01 January 1900 at noon and the time scale in which the Epoch was initialized. ([_Maybe_ hifitime should be even more precise?](https://github.com/nyx-space/hifitime/issues/186))
+
+## Importance of Kani in hifitime
+
+The purpose of hifitime is to convert between time scales, and these drift apart from each other. Time scales are only used for scientific computations, so it's important to correctly compute the conversion between these time scales. As previously discussed, for ensure exactly one nanosecond precision, hifitime stores durations as a tuple of integers, as these aren't affected by successive rounding errors. The drawback of this approach is that we now have to rewrite the arithmetics of durations and get it right!
```
Comment (tautschnig) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:40]: s/for ensure/to ensure/

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+This is where Kani comes in very handy. From its definition, Durations have a minimum and maximum value. On an unsigned 64 bit integer, we can store enough nanoseconds for four full centuries a bit. But we want want normalize all operations such that the nanosecond counter stays within one century (until we're reached the maximum number of centuries). Centuries are stored in a _signed_ 16 bit integer, and are the only field that store whether the duration is positive or negative.
```
Comment (tautschnig) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:42]: 1. Maybe that's just because English isn't my first language: what does "four full centuries a bit" mean?
2. s/want want/want to/
3. s/we're/we've/
4. s/that store/that stores/

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+Kani can also have unbounded tests where there is no explicit condition to check. Instead, only the successive operations of a function call are executed, and each are tested by Kani for failure cases by analyzing the inputs and finding cases where inputs will lead to runtime errors like overflows. This approach is how most of the bugs in hifitime have been found.
+
+Unbounded tests effectively ensure that sanity of the operations in a given function call. Bounded tests provide the same while also checking equalities. If either of these tests fail, Kani can provide a test failure report outlining the sequence of operations, and the binary representation of each intermediate operation, to help the developer gain an understanding of why their implementation is incorrect.
+
+The overhead to implement tests in Kani is very low, and the benefits are immense. Basically, write a Kani verification like a unit test, run the model verifier, and you've formally verified this part of the code. Amazing!
```
Comment (tautschnig) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:54]: We might want to reconsider the use of "tests" in these paragraphs.

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md (hunk 1 above)
Comment (remi-delmas-3000) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:16]: s/Earths does not in fact complete a full rotation/Earth does in fact complete a full rotation/

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+<img src="{{site.baseurl | prepend: site.url}}/assets/images/tai-vs-et-vs-tdb.png" alt="TT is fixed to TAI but TDB and ET are periodic" />
+
+
+## Avoiding loss of precision
+
+The precision of floating point values is constrained by its representation in the processor. Moreover, some microcontrollers do not have a floating point unit (FPU), meaning that any floating point operation is emulated through software, adding a considerable number of CPU cycles for any computation.
```
Comment (remi-delmas-3000) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:30]: The precision of a floating point value depends on its magnitude (as represented by the exponent) but also on how it was computed from other values (rounding modes, error accumulation, use of extended precision for intermediate computations in CPU, etc.)

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md (hunk 3 above)
Comment (remi-delmas-3000) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:40]: Can time scales be used for something else than scientific computation  ?

Comment (remi-delmas-3000) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:40]: We have to rewrite the arithmetics of durations on this (centuries, nanosecond) encoding and guarantee it satisfies expected properties.

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+This is where Kani comes in very handy. From its definition, Durations have a minimum and maximum value. On an unsigned 64 bit integer, we can store enough nanoseconds for four full centuries a bit. But we want want normalize all operations such that the nanosecond counter stays within one century (until we're reached the maximum number of centuries). Centuries are stored in a _signed_ 16 bit integer, and are the only field that store whether the duration is positive or negative.
+
+At first sight, it seems like a relatively simple task to make sure that this math is correct. It turns out that lots of bug may happen near the min and max durations, and when performing operations between very large and very small durations. Kani has helped fix at least eight different categories of bugs in a single pull request: <https://github.com/nyx-space/hifitime/pull/192>. Most of these were bugs near the boundaries of a Duration definition: around the zero, maximum, and minimum durations. But many of the bugs were on important and common operations: partial equality, negation, addition, and subtraction operations.
```
Comment (remi-delmas-3000) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:44]: suggestion : It turns out that handling edge cases near the min and max durations or when performing operations between very large and very small durations requires special attention.

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+One of the great things of Kani is that it doesn't require explicitly exhaustive tests. In fact, Kani operates on the mid-layer representation. It then solves for satifiability: "can this operation ever be executed, and if so, what are the cases where it may cause issues."
```
Comment (remi-delmas-3000) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:46]: Suggestion: "One of the great features of Kani is that it performs what is known as _symbolic execution_ of programs, where inputs are modelled as symbolic variables covering whole ranges of values at once. All program behaviours possible under these inputs are analysed for defects like arithmetic overflows or underflows, signed conversion overflow or underflow, etc.. If a defect is possible for some values of the inputs, Kani will generate a counter example trace with concrete values triggering the defect".

The mid-layer representation info does not seem that important to mention

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md (hunk 5 above)
Comment (remi-delmas-3000) [reply _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:54]: It seems like the way bounded/unbounded distinction is made here conflicts with the usual definitions.

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md (hunk 4 above)
Comment (ChristopherRabotin) [reply _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:42]: 1. Should be "four full centuries and a bit". On a one u64, we can store over four centuries of nanoseconds (but less than five centuries). Does that make sense?

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md (hunk 5 above)
Comment (ChristopherRabotin) [reply _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:54]: It's quite likely that I don't understand the difference between bounded and unbounded tests: can you explain the difference ?

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md (hunk 1 above)
Comment (ChristopherRabotin) [reply _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:16]: That's the point ;-) The 23 hours, 56 minutes and 4.091 seconds is an average duration of a sidereal day. In practice, this changes every day. In fact, that's what the UT1 time scale measures: the _exact_ duration it took for the Earth to perform a full rotation with respect to the stars.

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md (hunk 3 above)
Comment (ChristopherRabotin) [reply _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:40]: Using a glitch-free time scale is useful for a monotonic clock that can be synchronized with other systems. Apart from that and scientific computations that require these time scales, I'm not sure what other applications there are.

Comment (ChristopherRabotin) [reply _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md]: I've added some examples at the top of the article.

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md (hunk 1 above)
Comment (ChristopherRabotin) [reply _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md]: Clarified in the revision, let me know if it still needs work

File: _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+The figure below shows how the effect of gravity on the duration of a second compared to Earth based clocks, like TT (blue line fixed at 32.184 s) and TAI (reference Earth time, not shown since it would be a line at exactly zero).
+
+<img src="{{site.baseurl | prepend: site.url}}/assets/images/tai-vs-et-vs-tdb.png" alt="TT is fixed to TAI but TDB and ET are periodic" />
+
+
+## Avoiding loss of precision
+
+The precision of a floating point value depends on its magnitude (as represented by the exponent) and on how it was computed from other values (rounding modes, error accumulation, use of extended precision for intermediate computations in CPU, etc.). Moreover, some microcontrollers do not have a floating point unit (FPU), meaning that any floating point operation is emulated through software, adding a considerable number of CPU cycles for any computation.
+
+Dates, or "epochs," are _simply_ the duration past a certain reference epoch. Operations on these epochs, like adding 26.7 days, will cause rounding errors (even sometimes in a single operations like on this [`f32`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b760579f103b7192c20413ebbe167b90)).
```
Comment (zhassan-aws) [inline _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md:40]:
```suggestion
Dates, or "epochs," are _simply_ the duration past a certain reference epoch. Operations on these epochs, like adding 26.7 days, will cause rounding errors (even sometimes in a single operation like on this [`f32`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b760579f103b7192c20413ebbe167b90)).
```

File: _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+With Kani, we can check for all permutations of a definition of a `Duration`, and ensure that the decomposition of a `Duration` into its parts of days, hours, minutes, seconds, milliseconds, microseconds, and nanoseconds, never causes any overflow, underflow, or general unsoundness. In hifitime, this is done simply by implementing `Arbitrary` for `Duration`, and calling the `decompose()` function on _any_ duration. This tests is beautiful in its simplicity: small code footprint for mighty guarantees, such could be the motto of Kani!
+
+```rust
+#[cfg(kani)]
+impl Arbitrary for Duration {
```
Comment (zhassan-aws) [inline _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md:58]: A simpler alternative could be to `derive` `Arbitrary` directly, e.g.
```rust
use kani::Arbitrary;

#[cfg_attr(kani, derive(Arbitrary))]
pub struct Duration {
    pub(crate) centuries: i16,
    pub(crate) nanoseconds: u64,
}
```

File: _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+Kani can also have unbounded tests where there is no explicit condition to check. Instead, only the successive operations of a function call are executed, and each are tested by Kani for failure cases by analyzing the inputs and finding cases where inputs will lead to runtime errors like overflows. This approach is how most of the bugs in hifitime have been found.
+
+Unbounded tests effectively ensure that sanity of the operations in a given function call. Bounded tests provide the same while also checking equalities. If either of these tests fail, Kani can provide a test failure report outlining the sequence of operations, and the binary representation of each intermediate operation, to help the developer gain an understanding of why their implementation is incorrect.
+
+The overhead to implement tests in Kani is very low, and the benefits are immense. Basically, write a Kani verification like a unit test, run the model verifier, and you've formally verified this part of the code. Amazing!
```
Comment (zhassan-aws) [inline _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md:93]: Could be useful to mention the number of Kani tests that are integrated in `hifitime`.

**Review by zhassan-aws — COMMENTED** [review]
Looks great! A few comments.

File: _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+---
+
+Hifitime is a scientifically accurate time management library that provides nanosecond precision of durations and time computations for 65 thousand centuries in most time scales (also called time systems). It is useful wherever a monotonic clock is needed, even in the presence of leap second or remote time corrections (like for Network Time Protocol or Precise Time Protocol). It is suitable for desktop applications and for embedded systems without a floating point unit (FPU). It also supports the typical features of date time management libraries, like the [formatting and parsing of date times](https://docs.rs/hifitime/latest/hifitime/efmt/format/struct.Format.html) in a `no-std` environment using the typical C89 tokens, or a [human friendly approximation](https://docs.rs/hifitime/latest/hifitime/prelude/struct.Duration.html#method.approx) of the difference between dates. For scientific en engineering applications, it is mostly tailored to astronomy (with the inclusion of the UT1 time scale), astrodynamics (with the ET and TDB time scales), and global navigation satellite systems (with the GPS, Galileo, and BeiDou time scales).
```
[the remark this answers is not in the document]
Comment (ChristopherRabotin) [reply _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md:6]: Typo ;-) Will fix in the upcoming commit

**Comment by ChristopherRabotin** [comment]
I think I've implemented the review changes.

**Review by zhassan-aws — APPROVED** [review]

**Comment by rahulku** [comment]
@ChristopherRabotin - LMK if you have some more changes. I think including the link to the repo is useful.

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+This is where Kani comes in very handy. From its definition, Durations have a minimum and maximum value. On an unsigned 64 bit integer, we can store enough nanoseconds for four full centuries and a bit (but less than five centuries worth of nanoseconds). We want to normalize all operations such that the nanosecond counter stays within one century, unless we've reached the maximum (or minimum) number of centuries in which case, we want those nanoseconds to continue counting until the `u64` bound is reached. Centuries are stored in a _signed_ 16 bit integer, and are the only field that stores whether the duration is positive or negative.
```
Comment (feliperodri) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:42]: 64-bit, 16-bit...

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+At first sight, it seems like a relatively simple task to make sure that this math is correct. It turns out that handling edge cases near the min and max durations or when performing operations between very large and very small durations requires special attention. Kani has helped fix at least eight different categories of bugs in a single pull request: <https://github.com/nyx-space/hifitime/pull/192>. Most of these were bugs near the boundaries of a Duration definition: around the zero, maximum, and minimum durations. But many of the bugs were on important and common operations: partial equality, negation, addition, and subtraction operations. These bugs weren't due to lax testing: there are over 74 integration tests with plenty of checks within each.
```
Comment (feliperodri) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:44]: I think it'd be nice here to show some code of the bug with some values from the counterexample from Kani. (only a suggestion)

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+One of the great features of Kani is that it performs what is known as symbolic execution of programs, where inputs are modelled as symbolic variables covering whole ranges of values at once. All program behaviors possible under these inputs are analyzed for defects like arithmetic overflows or underflows, signed conversion overflow or underflow, etc.. If a defect is possible for some values of the inputs, Kani will generate a counter example trace with concrete values triggering the defect.
```
Comment (feliperodri) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:46]:
```suggestion
One of the great features of Kani is that it performs what is known as symbolic execution of programs, where inputs are modelled as symbolic variables covering whole ranges of values at once. All program behaviors possible under these inputs are analyzed for defects like arithmetic overflows or underflows, signed conversion overflow or underflow, etc. If a defect is possible for some values of the inputs, Kani will generate a counter example trace with concrete values triggering the defect.
```

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+Thanks to how Kani analyzes a program, tests can either be bounded or not. A bounded test includes an assertion: execute a set of instructions and then check something. This is a typical test case.
+
+Kani can also have unbounded tests where there is no explicit condition to check. Instead, only the successive operations of a function call are executed, and each are tested by Kani for failure cases by analyzing the inputs and finding cases where inputs will lead to runtime errors like overflows. This approach is how most of the bugs in hifitime have been found.
```
Comment (feliperodri) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:50]: Again, I think it'd be cool to see an example of a harness (just a suggestion).

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+Unbounded tests effectively ensure that sanity of the operations in a given function call. Bounded tests provide the same while also checking equalities. If either of these tests fail, Kani can provide a test failure report outlining the sequence of operations, and the binary representation of each intermediate operation, to help the developer gain an understanding of why their implementation is incorrect.
+
+The overhead to implement tests in Kani is very low, and the benefits are immense. Basically, write a Kani verification like a unit test, run the model verifier, and you've formally verified this part of the code. Amazing!
```
Comment (feliperodri) [inline _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:54]: I think it's really important here to highlight how assumptions are important as well. The verification result is as strong as your assumptions.

**Review by feliperodri — APPROVED** [review]
Looks great!! Only a few suggestions and minor corrections!

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md (hunk 17 above)
Comment (feliperodri) [reply _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md:54]: cc. @zhassan-aws @tautschnig @rahulku

File: _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md (hunk 14 above)
Comment (ChristopherRabotin) [reply _posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md]: I'm not sure which commit on hifitime would exhibit that erroneous behavior anymore. The updated blog post, that I guess I forgot to push last night, includes a specific edge case example as Rahul recommended.

**Comment by ChristopherRabotin** [comment]
I've implemented the following recommendations:

+ Add link to Beyond Measure
+ Add example test without explicit value checking
+ Write `16-bit` and `64-bit` with a dash
+ Specify that assumptions can be added to the test

File: _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+With Kani, we can check for all permutations of a definition of a `Duration`, and ensure that the decomposition of a `Duration` into its parts of days, hours, minutes, seconds, milliseconds, microseconds, and nanoseconds, never causes any overflow, underflow, or general unsoundness. In hifitime, this is done simply by implementing `Arbitrary` for `Duration`, and calling the `decompose()` function on _any_ duration. This tests is beautiful in its simplicity: small code footprint for mighty guarantees, such could be the motto of Kani!
```
Comment (feliperodri) [inline _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md:54]:
```suggestion
With Kani, we can check for all permutations of a definition of a `Duration`, and ensure that the decomposition of a `Duration` into its parts of days, hours, minutes, seconds, milliseconds, microseconds, and nanoseconds, never causes any overflow, underflow, or other undefined behaviors. In hifitime, this is done simply by implementing `Arbitrary` for `Duration`, and calling the `decompose()` function on _any_ duration. This test is beautiful in its simplicity: small code footprint for mighty guarantees, such could be the motto of Kani!
```

File: _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+```rust
+#[cfg(kani)]
+#[kani::proof]
+fn formal_duration_normalize_any() {
+    let dur: Duration = kani::any();
+    // Check that decompose never fails
+    dur.decompose();
+}
+```
```
Comment (feliperodri) [inline _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md:100]: You don't need to repeat this code, just point to the previous section.

**Review by feliperodri — APPROVED** [review]
Thank you for implementing all the recommendations. The blog post looks great!

**Comment by ChristopherRabotin** [comment]
Update ;-)

File: _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md
Comment (adpaco-aws) [inline _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md:1]: Can you rename this file to `_posts/2023-03-31-how-kani-helped-find-bugs-in-hifitime.md`?

File: _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+
+In the test above, if the call to `decompose` ever causes any overflow, underflow, division by zero, etc. then Kani will report an error and the exact value and bits inputted that caused the problem. This feedback is great, because one can just plugin those values into another test and debug it precisely. One will note that this test does not actually check the output value. That's for two reasons. First, testing values for all possible combinations would require re-implementing the same code that is in the `decompose` function making the test a tautology. Second, the purpose of the Hifitime Kani tests is to ensure that the aren't any unsound operations. Moreover, Hifitime has plenty of values that are explicitly tested for in the rest the tests.
```
Comment (remi-delmas-3000) [inline _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md:79]: > That's for two reasons. First, testing values for all possible combinations would require re-implementing the same code that is in the `decompose` function making the test a tautology.

This sentence makes it seem like there would be no value in checking the output of the operation and that tautology checking is not that useful. However formal verification is effectively tautology checking ! So I think the message could be confusing to some. 

Could we instead say something like : there would be some value in checking that the output of the operation is as expected, if we had written a declarative, formal specification of the input-output relation of `decompose`. However this was not a goal in this project." 

And maybe save that for future work?.

We could also equivalence check this function against a second and independently developed equivalent.

Comment (ChristopherRabotin) [reply _posts/2023-03-23-how-kani-helped-find-bugs-in-hifitime.md]: What do you think of the updated version?

> That's for two reasons. First, testing values would require developing either a formal specification or an independent implementation of the `decompose` function against which to check. Both options require quite a bit of effort, and the `decompose` function is really only used for displaying a duration in a human-readable format. Second, ...
